zerver/tests/test_unread.py

```diff
@@ -137,3 +137,109 @@ def test_update_all_flags(self):
+                                                    user_profile=user_profile
+        ).values_list('message__id',flat=True))
+
+        um = list(UserMessage.objects.filter(message__recipient=recipient, user_profile=user_profile))
+        for msg in um:
+            if msg.user_profile.email == "hamlet@zulip.com":
+                self.assertFalse(msg.flags.read)
+
+        events = []
+        with tornado_redirected_to_list(events):
+            result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                               "op": "add",
+                                                               "flag": "read",
+                                                               "stream_name": "Scotland"})
+        self.assert_json_success(result)
+
+        event = events[0]['event']
+        expected = dict(operation='add',
+                        messages=message_ids,
+                        flag='read',
+                        type='update_message_flags',
+                        all=False)
+        self.assertTrue(len([key for key in expected if expected[key] != event[key]]) == 0)
+
+        um = list(UserMessage.objects.filter(message__subject__iexact="Denmark3"))
+        for msg in um:
+            if msg.user_profile.email == "hamlet@zulip.com":
+                self.assertTrue(msg.flags.read)
+
+    def test_mark_all_in_invalid_stream_read(self):
+        self.login("hamlet@zulip.com")
+        invalid_stream_name = ""
+        result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                           "op": "add",
+                                                           "flag": "read",
+                                                           "stream_name": invalid_stream_name})
+        self.assert_json_error(result, 'No such stream \'\'')
+
+    def test_mark_all_in_stream_topic_read(self):
+        self.login("hamlet@zulip.com")
+
+        result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                           "op": "remove",
+                                                           "flag": "read",
+                                                           "topic_name": "Denmark3",
+                                                           "stream_name": "Denmark"})
+
+        self.assert_json_success(result)
+        user_profile = get_user_profile_by_email("hamlet@zulip.com")
+        stream = get_stream("Denmark", user_profile.realm)
+        recipient = get_recipient(Recipient.STREAM, stream.id)
+        message_ids = list(UserMessage.objects.filter(message__recipient=recipient,
+                                                    user_profile=user_profile,
+                                                    message__subject__iexact="Denmark3"
+        ).values_list('message__id',flat=True))
```

Comment:
oh, sorry, my comment was confusing, because the closest query to it was an `update_message_flags` query.  This was meant to be about the result object you get from a `send_message` query (e.g. on line 148), not an `update_message_flags` query.